File: app/src/main/java/com/veganbeauty/app/features/shop/product/CartBottomSheetFragment.java
package com.veganbeauty.app.features.shop.product;

import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LifecycleOwnerKt;
import androidx.lifecycle.FlowLiveDataConversions;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.veganbeauty.app.R;
import com.veganbeauty.app.data.local.ProfileSession;
import com.veganbeauty.app.data.local.RootieDatabase;
import com.veganbeauty.app.data.local.entities.CartItemEntity;
import com.veganbeauty.app.databinding.ShopBottomSheetCartBinding;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import kotlinx.coroutines.BuildersKt;
import kotlinx.coroutines.Dispatchers;

public class CartBottomSheetFragment extends BottomSheetDialogFragment {

    public static final String TAG = "CartBottomSheetFragment";
    public static final String ARG_SELECTED_VOUCHER_CODE = "selected_voucher_code";
    public static final String ARG_SELECTED_VOUCHER_DISCOUNT = "selected_voucher_discount";

    private ShopBottomSheetCartBinding _binding;
    private RootieDatabase database;
    private CartAdapter adapter;

    private String selectedVoucherCode = null;
    private long voucherDiscountAmount = 0L;
    private boolean isPointsChecked = false;

    public static CartBottomSheetFragment newInstance(String selectedCode, long discount) {
        CartBottomSheetFragment fragment = new CartBottomSheetFragment();
        Bundle args = new Bundle();
        args.putString(ARG_SELECTED_VOUCHER_CODE, selectedCode);
        args.putLong(ARG_SELECTED_VOUCHER_DISCOUNT, discount);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            selectedVoucherCode = getArguments().getString(ARG_SELECTED_VOUCHER_CODE);
            voucherDiscountAmount = getArguments().getLong(ARG_SELECTED_VOUCHER_DISCOUNT, 0L);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        _binding = ShopBottomSheetCartBinding.inflate(inflater, container, false);
        database = RootieDatabase.getDatabase(requireContext());
        return _binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (_binding == null) return;

        setupRecyclerView();
        setupListeners();
        applyGuestCartMode();
        observeCartItems();
    }

    private void applyGuestCartMode() {
        if (!ProfileSession.isLoggedIn(requireContext())) {
            _binding.llPointsRow.setVisibility(View.GONE);
            isPointsChecked = false;
            return;
        }
        _binding.llPointsRow.setVisibility(View.VISIBLE);
        refreshPointsRow();
    }

    private void setupRecyclerView() {
        adapter = new CartAdapter(
                (item, newQuantity) -> {
                    new Thread(() -> {
                        if (newQuantity <= 0) {
                            database.cartDao().deleteCartItem(item);
                            if (getActivity() != null) {
                                getActivity().runOnUiThread(() -> {
                                    Toast.makeText(requireContext(), "Đã xóa " + item.getName() + " khỏi giỏ hàng", Toast.LENGTH_SHORT).show();
                                });
                            }
                        } else {
                            CartItemEntity updatedItem = new CartItemEntity(
                                    item.getId(), item.getName(), item.getImage(), item.getPrice(), newQuantity, item.isSelected()
                            );
                            database.cartDao().updateCartItem(updatedItem);
                        }
                    }).start();
                },
                (item, isSelected) -> {
                    new Thread(() -> {
                        CartItemEntity updatedItem = new CartItemEntity(
                                item.getId(), item.getName(), item.getImage(), item.getPrice(), item.getQuantity(), isSelected
                        );
                        database.cartDao().updateCartItem(updatedItem);
                    }).start();
                }
        );

        _binding.rvCartItems.setLayoutManager(new LinearLayoutManager(requireContext()));
        _binding.rvCartItems.setAdapter(adapter);
    }

    private void setupListeners() {
        _binding.llSelectAll.setOnClickListener(v -> {
            List<CartItemEntity> currentList = adapter.getCurrentList();
            boolean allSelected = true;
            for (CartItemEntity item : currentList) {
                if (!item.isSelected()) {
                    allSelected = false;
                    break;
                }
            }
            boolean finalAllSelected = allSelected;
            new Thread(() -> {
                for (CartItemEntity item : currentList) {
                    CartItemEntity updatedItem = new CartItemEntity(
                            item.getId(), item.getName(), item.getImage(), item.getPrice(), item.getQuantity(), !finalAllSelected
                    );
                    database.cartDao().updateCartItem(updatedItem);
                }
            }).start();
        });

        _binding.switchPoints.setOnClickListener(v -> refreshPointsRow());

        _binding.llVoucherRow.setOnClickListener(v -> {
            ShopVoucherFragment voucherFragment = ShopVoucherFragment.newInstance(selectedVoucherCode);
            requireActivity().getSupportFragmentManager().beginTransaction()
                    .replace(R.id.main_container, voucherFragment)
                    .addToBackStack(null)
                    .commit();
            dismiss();
        });

        _binding.llPrices.setOnClickListener(v -> {
            boolean isVisible = _binding.llPriceDetails.getVisibility() == View.VISIBLE;
            if (isVisible) {
                _binding.llPriceDetails.setVisibility(View.GONE);
                _binding.ivPriceToggleArrow.setRotation(270f);
            } else {
                _binding.llPriceDetails.setVisibility(View.VISIBLE);
                _binding.ivPriceToggleArrow.setRotation(90f);

                if (getDialog() instanceof BottomSheetDialog) {
                    BottomSheetDialog bottomSheetDialog = (BottomSheetDialog) getDialog();
                    View bottomSheet = bottomSheetDialog.findViewById(com.google.android.material.R.id.design_bottom_sheet);
                    if (bottomSheet != null) {
                        BottomSheetBehavior<View> behavior = BottomSheetBehavior.from(bottomSheet);
                        behavior.setState(BottomSheetBehavior.STATE_EXPANDED);
                    }
                }
            }
        });

        _binding.btnCheckout.setOnClickListener(v -> {
            List<CartItemEntity> selectedItems = new ArrayList<>();
            for (CartItemEntity item : adapter.getCurrentList()) {
                if (item.isSelected()) {
                    selectedItems.add(item);
                }
            }

            if (selectedItems.isEmpty()) {
                Toast.makeText(requireContext(), "Vui lòng chọn ít nhất 1 sản phẩm để thanh toán", Toast.LENGTH_SHORT).show();
            } else {
                ArrayList<CartItemEntity> checkoutItems = new ArrayList<>(selectedItems);
                ShopCheckoutFragment checkoutFragment = ShopCheckoutFragment.newInstance(
                        checkoutItems,
                        selectedVoucherCode,
                        voucherDiscountAmount
                );
                requireActivity().getSupportFragmentManager().beginTransaction()
                        .replace(R.id.main_container, checkoutFragment)
                        .addToBackStack(null)
                        .commit();
                dismiss();
            }
        });
    }

    private void observeCartItems() {
        FlowLiveDataConversions.asLiveData(database.cartDao().getAllCartItems())
            .observe(getViewLifecycleOwner(), items -> {
                if (items != null) {
                    adapter.submitList(items);
                    updateUI(items);
                }
            });
    }

    private void updateUI(List<CartItemEntity> items) {
        NumberFormat formatter = NumberFormat.getCurrencyInstance(new Locale("vi", "VN"));

        List<CartItemEntity> selectedItems = new ArrayList<>();
        int totalSelectedQty = 0;
        for (CartItemEntity item : items) {
            if (item.isSelected()) {
                selectedItems.add(item);
                totalSelectedQty += item.getQuantity();
            }
        }

        long originalPriceSum = 0;
        long finalPriceSum = 0;
        for (CartItemEntity item : selectedItems) {
            originalPriceSum += (long) (item.getPrice() * 1.2) * item.getQuantity();
            finalPriceSum += item.getPrice() * item.getQuantity();
        }

        boolean isVoucherApplied = selectedVoucherCode != null && !selectedVoucherCode.isEmpty();
        if (isVoucherApplied && finalPriceSum > voucherDiscountAmount) {
            finalPriceSum -= voucherDiscountAmount;
        }

        long totalSavings = originalPriceSum - finalPriceSum;

        _binding.tvDetailSubtotal.setText(formatter.format(originalPriceSum));
        long directDiscount = originalPriceSum - finalPriceSum;
        if (directDiscount - voucherDiscountAmount > 0) {
            _binding.tvDetailDirectDiscount.setText("-" + formatter.format(directDiscount - voucherDiscountAmount));
        } else {
            _binding.tvDetailDirectDiscount.setText("0đ");
        }

        if (voucherDiscountAmount > 0) {
            _binding.tvDetailVoucherDiscount.setText("-" + formatter.format(voucherDiscountAmount));
        } else {
            _binding.tvDetailVoucherDiscount.setText("0đ");
        }

        if (isVoucherApplied) {
            _binding.tvVoucherDesc.setText("Đã áp dụng mã: " + selectedVoucherCode + " (-" + formatter.format(voucherDiscountAmount) + ")");
            _binding.llVoucherRow.setBackgroundTintList(ColorStateList.valueOf(Color.parseColor("#E5F2FF")));
        } else {
            _binding.tvVoucherDesc.setText("Áp dụng ưu đãi để được giảm giá");
            _binding.llVoucherRow.setBackgroundTintList(ColorStateList.valueOf(Color.parseColor("#EDF3ED")));
        }

        _binding.tvOriginalTotalPrice.setText(formatter.format(originalPriceSum));
        _binding.tvOriginalTotalPrice.setPaintFlags(_binding.tvOriginalTotalPrice.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);

        _binding.tvTotalValue.setText(formatter.format(finalPriceSum));
        _binding.tvSavingsValue.setText(formatter.format(totalSavings));

        refreshPointsRow();

        _binding.btnCheckout.setText("Mua hàng");

        _binding.tvSelectAllLabel.setText("Chọn tất cả (" + items.size() + ")");
        boolean allSelected = !items.isEmpty();
        for (CartItemEntity item : items) {
            if (!item.isSelected()) {
                allSelected = false;
                break;
            }
        }
        if (allSelected) {
            _binding.ivSelectAll.setImageResource(R.drawable.ic_circle_checked);
            _binding.ivSelectAll.setImageTintList(ColorStateList.valueOf(Color.parseColor("#3E4D44")));
        } else {
            _binding.ivSelectAll.setImageResource(R.drawable.ic_circle);
            _binding.ivSelectAll.setImageTintList(null);
        }
    }

    private void refreshPointsRow() {
        if (_binding == null || !isAdded() || !ProfileSession.isLoggedIn(requireContext())) return;
        int balance = Math.max(0, com.veganbeauty.app.utils.RewardPointsHelper.getTotalPoints(requireContext()));
        String balanceText = String.format(Locale.getDefault(), "%,d", balance).replace(',', '.');
        _binding.tvPointsText.setText("Bạn có " + balanceText + " xu");
        isPointsChecked = false;
        updateSwitchUI(_binding.switchPoints, _binding.switchPointsThumb, false);
    }

    private void updateSwitchUI(FrameLayout container, ImageView thumb, boolean enabled) {
        if (enabled) {
            container.setBackgroundResource(R.drawable.ic_switch_track_on);
            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) thumb.getLayoutParams();
            lp.gravity = android.view.Gravity.CENTER_VERTICAL | android.view.Gravity.END;
            lp.setMarginStart(0);
            lp.setMarginEnd((int) (2 * getResources().getDisplayMetrics().density));
            thumb.setLayoutParams(lp);
        } else {
            container.setBackgroundResource(R.drawable.ic_switch_track_off);
            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) thumb.getLayoutParams();
            lp.gravity = android.view.Gravity.CENTER_VERTICAL | android.view.Gravity.START;
            lp.setMarginEnd(0);
            lp.setMarginStart((int) (2 * getResources().getDisplayMetrics().density));
            thumb.setLayoutParams(lp);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        _binding = null;
    }
}

File: app/src/main/java/com/veganbeauty/app/features/shop/product/CartBottomSheetFragment.kt
package com.veganbeauty.app.features.shop.product

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.veganbeauty.app.R
import com.veganbeauty.app.data.local.RootieDatabase
import com.veganbeauty.app.data.local.entities.CartItemEntity
import com.veganbeauty.app.databinding.ShopBottomSheetCartBinding
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.text.NumberFormat
import java.util.Locale

class CartBottomSheetFragment : BottomSheetDialogFragment() {

    private var _binding: ShopBottomSheetCartBinding? = null
    private val binding get() = _binding!!

    private lateinit var database: RootieDatabase
    private lateinit var adapter: CartAdapter

    private var selectedVoucherCode: String? = null
    private var voucherDiscountAmount = 0L
    private val isVoucherApplied get() = !selectedVoucherCode.isNullOrEmpty()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            selectedVoucherCode = it.getString(ARG_SELECTED_VOUCHER_CODE)
            voucherDiscountAmount = it.getLong(ARG_SELECTED_VOUCHER_DISCOUNT, 0L)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        if (!com.veganbeauty.app.data.local.ProfileSession.isLoggedIn(requireContext())) {
            Toast.makeText(requireContext(), "Vui lòng đăng nhập để xem giỏ hàng", Toast.LENGTH_SHORT).show()
            dismiss()
            return null
        }
        _binding = ShopBottomSheetCartBinding.inflate(inflater, container, false)
        database = RootieDatabase.getDatabase(requireContext())
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (_binding == null) return

        setupRecyclerView()
        setupListeners()
        observeCartItems()
    }

    private fun setupRecyclerView() {
        adapter = CartAdapter(
            onQuantityChanged = { item, newQuantity ->
                lifecycleScope.launch {
                    if (newQuantity <= 0) {
                        database.cartDao().deleteCartItem(item)
                        Toast.makeText(requireContext(), "Đã xóa ${item.name} khỏi giỏ hàng", Toast.LENGTH_SHORT).show()
                    } else {
                        database.cartDao().updateCartItem(item.copy(quantity = newQuantity))
                    }
                }
            },
            onSelectionToggled = { item, isSelected ->
                lifecycleScope.launch {
                    database.cartDao().updateCartItem(item.copy(isSelected = isSelected))
                }
            }
        )

        binding.rvCartItems.layoutManager = LinearLayoutManager(requireContext())
        binding.rvCartItems.adapter = adapter
    }

    private fun setupListeners() {
        // Toggle Select All clicked
        binding.llSelectAll.setOnClickListener {
            val currentList = adapter.currentList
            val allSelected = currentList.all { it.isSelected }
            lifecycleScope.launch {
                currentList.forEach { item ->
                    database.cartDao().updateCartItem(item.copy(isSelected = !allSelected))
                }
            }
        }

        // Points Switch Toggle
        binding.switchPoints.setOnCheckedChangeListener { _, _ ->
            updateUI(adapter.currentList)
        }

        // Voucher Row Click
        binding.llVoucherRow.setOnClickListener {
            val voucherFragment = ShopVoucherFragment.newInstance(selectedVoucherCode)
            parentFragmentManager.beginTransaction()
                .replace(R.id.main_container, voucherFragment)
                .addToBackStack(null)
                .commit()
            dismiss()
        }

        // Toggle Price details click
        binding.llPrices.setOnClickListener {
            val isVisible = binding.llPriceDetails.visibility == View.VISIBLE
            if (isVisible) {
                binding.llPriceDetails.visibility = View.GONE
                binding.ivPriceToggleArrow.rotation = 270f
            } else {
                binding.llPriceDetails.visibility = View.VISIBLE
                binding.ivPriceToggleArrow.rotation = 90f
                
                // Force bottom sheet behavior to expand fully so footer remains visible
                (dialog as? com.google.android.material.bottomsheet.BottomSheetDialog)?.let { bottomSheetDialog ->
                    val bottomSheet = bottomSheetDialog.findViewById<View>(com.google.android.material.R.id.design_bottom_sheet)
                    if (bottomSheet != null) {
                        val behavior = com.google.android.material.bottomsheet.BottomSheetBehavior.from(bottomSheet)
                        behavior.state = com.google.android.material.bottomsheet.BottomSheetBehavior.STATE_EXPANDED
                    }
                }
            }
        }

        // Checkout button
        binding.btnCheckout.setOnClickListener {
            val selectedItems = adapter.currentList.filter { it.isSelected }
            if (selectedItems.isEmpty()) {
                Toast.makeText(requireContext(), "Vui lòng chọn ít nhất 1 sản phẩm để thanh toán", Toast.LENGTH_SHORT).show()
            } else {
                val checkoutItems = ArrayList(selectedItems)
                val checkoutFragment = ShopCheckoutFragment.newInstance(
                    checkoutItems,
                    selectedVoucherCode,
                    voucherDiscountAmount
                )
                parentFragmentManager.beginTransaction()
                    .replace(R.id.main_container, checkoutFragment)
                    .addToBackStack(null)
                    .commit()
                dismiss()
            }
        }
    }

    private fun observeCartItems() {
        lifecycleScope.launch {
            database.cartDao().getAllCartItems().collectLatest { items ->
                adapter.submitList(items)
                updateUI(items)
            }
        }
    }

    private fun updateUI(items: List<CartItemEntity>) {
        val formatter = NumberFormat.getCurrencyInstance(Locale("vi", "VN"))

        // Calculate count of selected items
        val selectedItems = items.filter { it.isSelected }
        val totalSelectedQty = selectedItems.sumOf { it.quantity }
        
        val originalPriceSum = selectedItems.sumOf { (it.price * 1.2).toLong() * it.quantity }
        var finalPriceSum = selectedItems.sumOf { it.price * it.quantity }

        // Points deduction
        if (binding.switchPoints.isChecked && finalPriceSum > 2400) {
            finalPriceSum -= 2400
        }

        // Voucher deduction
        if (isVoucherApplied && finalPriceSum > voucherDiscountAmount) {
            finalPriceSum -= voucherDiscountAmount
        }

        val totalSavings = originalPriceSum - finalPriceSum

        // Update details values
        binding.tvDetailSubtotal.text = formatter.format(originalPriceSum)
        val directDiscount = originalPriceSum - finalPriceSum
        binding.tvDetailDirectDiscount.text = if (directDiscount - voucherDiscountAmount > 0) {
            "-${formatter.format(directDiscount - voucherDiscountAmount)}"
        } else {
            "0đ"
        }
        
        binding.tvDetailVoucherDiscount.text = if (voucherDiscountAmount > 0) {
            "-${formatter.format(voucherDiscountAmount)}"
        } else {
            "0đ"
        }

        // Update Voucher row visual state
        if (isVoucherApplied) {
            binding.tvVoucherDesc.text = "Đã áp dụng mã: $selectedVoucherCode (-${formatter.format(voucherDiscountAmount)})"
            binding.llVoucherRow.setBackgroundTintList(android.content.res.ColorStateList.valueOf(android.graphics.Color.parseColor("#E5F2FF")))
        } else {
            binding.tvVoucherDesc.text = "Áp dụng ưu đãi để được giảm giá"
            binding.llVoucherRow.setBackgroundTintList(android.content.res.ColorStateList.valueOf(android.graphics.Color.parseColor("#EDF3ED")))
        }

        // Update total values
        binding.tvOriginalTotalPrice.text = formatter.format(originalPriceSum)
        binding.tvOriginalTotalPrice.paintFlags = binding.tvOriginalTotalPrice.paintFlags or android.graphics.Paint.STRIKE_THRU_TEXT_FLAG
        
        binding.tvTotalValue.text = formatter.format(finalPriceSum)
        binding.tvSavingsValue.text = formatter.format(totalSavings)

        binding.btnCheckout.text = "Mua hàng"

        // Update select all text and icon state
        binding.tvSelectAllLabel.text = "Chọn tất cả (${items.size})"
        val allSelected = items.isNotEmpty() && items.all { it.isSelected }
        if (allSelected) {
            binding.ivSelectAll.setImageResource(R.drawable.ic_cart_checked)
        } else {
            binding.ivSelectAll.setImageResource(R.drawable.ic_cart_unchecked)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        const val TAG = "CartBottomSheetFragment"
        const val ARG_SELECTED_VOUCHER_CODE = "selected_voucher_code"
        const val ARG_SELECTED_VOUCHER_DISCOUNT = "selected_voucher_discount"

        fun newInstance(selectedCode: String?, discount: Long): CartBottomSheetFragment {
            return CartBottomSheetFragment().apply {
                arguments = Bundle().apply {
                    putString(ARG_SELECTED_VOUCHER_CODE, selectedCode)
                    putLong(ARG_SELECTED_VOUCHER_DISCOUNT, discount)
                }
            }
        }
    }
}
